File: packages/firebase_common/lib/src/firestore_documents/user_fcm_token.flutterfire_gen.dart
// coverage:ignore-file
// ignore_for_file: type=lint
// ignore_for_file: unused_element, deprecated_member_use, deprecated_member_use_from_same_package, use_function_type_syntax_for_parameters, unnecessary_const, avoid_init_to_null, invalid_override_different_default_values_named, prefer_expression_function_bodies, annotate_overrides, invalid_annotation_target, unnecessary_question_mark

part of 'user_fcm_token.dart';

class ReadUserFcmToken {
  const ReadUserFcmToken({
    required this.userFcmTokenId,
    required this.path,
    required this.userId,
    required this.token,
    required this.deviceInfo,
    required this.createdAt,
    required this.updatedAt,
  });

  final String userFcmTokenId;

  final String path;

  final String userId;

  final String token;

  final String deviceInfo;

  final DateTime? createdAt;

  final DateTime? updatedAt;

  factory ReadUserFcmToken._fromJson(Map<String, dynamic> json) {
    return ReadUserFcmToken(
      userFcmTokenId: json['userFcmTokenId'] as String,
      path: json['path'] as String,
      userId: json['userId'] as String,
      token: json['token'] as String,
      deviceInfo: json['deviceInfo'] as String? ?? '',
      createdAt: (json['createdAt'] as Timestamp?)?.toDate(),
      updatedAt: (json['updatedAt'] as Timestamp?)?.toDate(),
    );
  }

  factory ReadUserFcmToken.fromDocumentSnapshot(DocumentSnapshot ds) {
    final data = ds.data()! as Map<String, dynamic>;
    return ReadUserFcmToken._fromJson(<String, dynamic>{
      ...data,
      'userFcmTokenId': ds.id,
      'path': ds.reference.path,
    });
  }
}

class CreateUserFcmToken {
  const CreateUserFcmToken({
    required this.userId,
    required this.token,
    required this.deviceInfo,
  });

  final String userId;
  final String token;
  final String deviceInfo;

  Map<String, dynamic> toJson() {
    return {
      'userId': userId,
      'token': token,
      'deviceInfo': deviceInfo,
      'createdAt': FieldValue.serverTimestamp(),
      'updatedAt': FieldValue.serverTimestamp(),
    };
  }
}

class UpdateUserFcmToken {
  const UpdateUserFcmToken({
    this.userId,
    this.token,
    this.deviceInfo,
    this.createdAt,
  });

  final String? userId;
  final String? token;
  final String? deviceInfo;
  final DateTime? createdAt;

  Map<String, dynamic> toJson() {
    return {
      if (userId != null) 'userId': userId,
      if (token != null) 'token': token,
      if (deviceInfo != null) 'deviceInfo': deviceInfo,
      if (createdAt != null) 'createdAt': createdAt,
      'updatedAt': FieldValue.serverTimestamp(),
    };
  }
}

class DeleteUserFcmToken {}

/// Provides a reference to the userFcmTokens collection for reading.
final readUserFcmTokenCollectionReference = FirebaseFirestore.instance
    .collection('userFcmTokens')
    .withConverter<ReadUserFcmToken>(
      fromFirestore: (ds, _) => ReadUserFcmToken.fromDocumentSnapshot(ds),
      toFirestore: (_, __) => throw UnimplementedError(),
    );

/// Provides a reference to a userFcmToken document for reading.
DocumentReference<ReadUserFcmToken> readUserFcmTokenDocumentReference({
  required String userFcmTokenId,
}) =>
    readUserFcmTokenCollectionReference.doc(userFcmTokenId);

/// Provides a reference to the userFcmTokens collection for creating.
final createUserFcmTokenCollectionReference = FirebaseFirestore.instance
    .collection('userFcmTokens')
    .withConverter<CreateUserFcmToken>(
      fromFirestore: (_, __) => throw UnimplementedError(),
      toFirestore: (obj, _) => obj.toJson(),
    );

/// Provides a reference to a userFcmToken document for creating.
DocumentReference<CreateUserFcmToken> createUserFcmTokenDocumentReference({
  required String userFcmTokenId,
}) =>
    createUserFcmTokenCollectionReference.doc(userFcmTokenId);

/// Provides a reference to the userFcmTokens collection for updating.
final updateUserFcmTokenCollectionReference = FirebaseFirestore.instance
    .collection('userFcmTokens')
    .withConverter<UpdateUserFcmToken>(
      fromFirestore: (_, __) => throw UnimplementedError(),
      toFirestore: (obj, _) => obj.toJson(),
    );

/// Provides a reference to a userFcmToken document for updating.
DocumentReference<UpdateUserFcmToken> updateUserFcmTokenDocumentReference({
  required String userFcmTokenId,
}) =>
    updateUserFcmTokenCollectionReference.doc(userFcmTokenId);

/// Provides a reference to the userFcmTokens collection for deleting.
final deleteUserFcmTokenCollectionReference = FirebaseFirestore.instance
    .collection('userFcmTokens')
    .withConverter<DeleteUserFcmToken>(
      fromFirestore: (_, __) => throw UnimplementedError(),
      toFirestore: (_, __) => throw UnimplementedError(),
    );

/// Provides a reference to a userFcmToken document for deleting.
DocumentReference<DeleteUserFcmToken> deleteUserFcmTokenDocumentReference({
  required String userFcmTokenId,
}) =>
    deleteUserFcmTokenCollectionReference.doc(userFcmTokenId);

/// Manages queries against the userFcmTokens collection.
class UserFcmTokenQuery {
  /// Fetches [ReadUserFcmToken] documents.
  Future<List<ReadUserFcmToken>> fetchDocuments({
    GetOptions? options,
    Query<ReadUserFcmToken>? Function(Query<ReadUserFcmToken> query)?
        queryBuilder,
    int Function(ReadUserFcmToken lhs, ReadUserFcmToken rhs)? compare,
  }) async {
    Query<ReadUserFcmToken> query = readUserFcmTokenCollectionReference;
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    final qs = await query.get(options);
    final result = qs.docs.map((qds) => qds.data()).toList();
    if (compare != null) {
      result.sort(compare);
    }
    return result;
  }

  /// Subscribes [UserFcmToken] documents.
  Stream<List<ReadUserFcmToken>> subscribeDocuments({
    Query<ReadUserFcmToken>? Function(Query<ReadUserFcmToken> query)?
        queryBuilder,
    int Function(ReadUserFcmToken lhs, ReadUserFcmToken rhs)? compare,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    Query<ReadUserFcmToken> query = readUserFcmTokenCollectionReference;
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    var streamQs =
        query.snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamQs = streamQs.where((qs) => !qs.metadata.hasPendingWrites);
    }
    return streamQs.map((qs) {
      final result = qs.docs.map((qds) => qds.data()).toList();
      if (compare != null) {
        result.sort(compare);
      }
      return result;
    });
  }

  /// Fetches a specific [ReadUserFcmToken] document.
  Future<ReadUserFcmToken?> fetchDocument({
    required String userFcmTokenId,
    GetOptions? options,
  }) async {
    final ds = await readUserFcmTokenDocumentReference(
      userFcmTokenId: userFcmTokenId,
    ).get(options);
    return ds.data();
  }

  /// Subscribes a specific [UserFcmToken] document.
  Stream<ReadUserFcmToken?> subscribeDocument({
    required String userFcmTokenId,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    var streamDs = readUserFcmTokenDocumentReference(
      userFcmTokenId: userFcmTokenId,
    ).snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamDs = streamDs.where((ds) => !ds.metadata.hasPendingWrites);
    }
    return streamDs.map((ds) => ds.data());
  }

  /// Adds a [UserFcmToken] document.
  Future<DocumentReference<CreateUserFcmToken>> add({
    required CreateUserFcmToken createUserFcmToken,
  }) =>
      createUserFcmTokenCollectionReference.add(createUserFcmToken);

  /// Sets a [UserFcmToken] document.
  Future<void> set({
    required String userFcmTokenId,
    required CreateUserFcmToken createUserFcmToken,
    SetOptions? options,
  }) =>
      createUserFcmTokenDocumentReference(
        userFcmTokenId: userFcmTokenId,
      ).set(createUserFcmToken, options);

  /// Updates a specific [UserFcmToken] document.
  Future<void> update({
    required String userFcmTokenId,
    required UpdateUserFcmToken updateUserFcmToken,
  }) =>
      updateUserFcmTokenDocumentReference(
        userFcmTokenId: userFcmTokenId,
      ).update(updateUserFcmToken.toJson());

  /// Deletes a specific [UserFcmToken] document.
  Future<void> delete({
    required String userFcmTokenId,
  }) =>
      deleteUserFcmTokenDocumentReference(
        userFcmTokenId: userFcmTokenId,
      ).delete();
}

File: functions/src/firestore-documents/userFcmToken.flutterfireGen.ts
import * as admin from 'firebase-admin'
import {
    DocumentData,
    DocumentReference,
    DocumentSnapshot,
    FieldValue,
    Query,
    QueryDocumentSnapshot,
    QuerySnapshot,
    SetOptions,
    Timestamp,
    WriteResult
} from 'firebase-admin/firestore'

export class ReadUserFcmToken {
    constructor({
        userFcmTokenId,
        path,
        userId,
        token,
        deviceInfo,
        createdAt,
        updatedAt
    }: {
        userFcmTokenId: string
        path: string
        userId: string
        token: string
        deviceInfo: string
        createdAt?: Date
        updatedAt?: Date
    }) {
        this.userFcmTokenId = userFcmTokenId
        this.path = path
        this.userId = userId
        this.token = token
        this.deviceInfo = deviceInfo
        this.createdAt = createdAt
        this.updatedAt = updatedAt
    }

    readonly userFcmTokenId: string

    readonly path: string

    readonly userId: string

    readonly token: string

    readonly deviceInfo: string

    readonly createdAt?: Date

    readonly updatedAt?: Date

    private static fromJson(json: Record<string, unknown>): ReadUserFcmToken {
        return new ReadUserFcmToken({
            userFcmTokenId: json[`userFcmTokenId`] as string,
            path: json[`path`] as string,
            userId: json[`userId`] as string,
            token: json[`token`] as string,
            deviceInfo: (json[`deviceInfo`] as string | undefined) ?? ``,
            createdAt: (json[`createdAt`] as Timestamp | undefined)?.toDate(),
            updatedAt: (json[`updatedAt`] as Timestamp | undefined)?.toDate()
        })
    }

    static fromDocumentSnapshot(ds: DocumentSnapshot): ReadUserFcmToken {
        const data = ds.data()!
        const cleanedData: Record<string, unknown> = {}
        for (const [key, value] of Object.entries(data)) {
            cleanedData[key] = value === null ? undefined : value
        }
        return ReadUserFcmToken.fromJson({
            ...cleanedData,
            userFcmTokenId: ds.id,
            path: ds.ref.path
        })
    }
}

export class CreateUserFcmToken {
    constructor({
        userId,
        token,
        deviceInfo
    }: {
        userId: string
        token: string
        deviceInfo: string
    }) {
        this.userId = userId
        this.token = token
        this.deviceInfo = deviceInfo
    }

    readonly userId: string

    readonly token: string

    readonly deviceInfo: string

    toJson(): Record<string, unknown> {
        return {
            userId: this.userId,
            token: this.token,
            deviceInfo: this.deviceInfo,
            createdAt: FieldValue.serverTimestamp(),
            updatedAt: FieldValue.serverTimestamp()
        }
    }
}

export class UpdateUserFcmToken {
    constructor({
        userId,
        token,
        deviceInfo,
        createdAt
    }: {
        userId?: string
        token?: string
        deviceInfo?: string
        createdAt?: Date
    }) {
        this.userId = userId
        this.token = token
        this.deviceInfo = deviceInfo
        this.createdAt = createdAt
    }

    readonly userId?: string

    readonly token?: string

    readonly deviceInfo?: string

    readonly createdAt?: Date

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    toJson(): Record<string, any> {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const json: Record<string, any> = {}
        if (this.userId != undefined) {
            json[`userId`] = this.userId
        }
        if (this.token != undefined) {
            json[`token`] = this.token
        }
        if (this.deviceInfo != undefined) {
            json[`deviceInfo`] = this.deviceInfo
        }
        if (this.createdAt != undefined) {
            json[`createdAt`] = this.createdAt
        }
        json[`updatedAt`] = FieldValue.serverTimestamp()
        return json
    }
}

export class DeleteUserFcmToken {}

/**
 * A Cloud Firestore object which ignores `undefined` properties.
 */
const db = admin.firestore()
db.settings({ ignoreUndefinedProperties: true })

/**
 * Provides a reference to the userFcmTokens collection for reading.
 */
export const readUserFcmTokenCollectionReference = db
    .collection(`userFcmTokens`)
    .withConverter<ReadUserFcmToken>({
        fromFirestore: (ds: DocumentSnapshot): ReadUserFcmToken => {
            return ReadUserFcmToken.fromDocumentSnapshot(ds)
        },
        toFirestore: () => {
            throw new Error(
                `toFirestore is not implemented for ReadUserFcmToken`
            )
        }
    })

/**
 * Provides a reference to a userFcmToken document for reading.
 * @param userFcmTokenId - The ID of the userFcmToken document to read.
 */
export const readUserFcmTokenDocumentReference = ({
    userFcmTokenId
}: {
    userFcmTokenId: string
}): DocumentReference<ReadUserFcmToken> =>
    readUserFcmTokenCollectionReference.doc(userFcmTokenId)

/**
 * Provides a reference to the userFcmTokens collection for creating.
 */
export const createUserFcmTokenCollectionReference = db
    .collection(`userFcmTokens`)
    .withConverter<CreateUserFcmToken>({
        fromFirestore: () => {
            throw new Error(
                `fromFirestore is not implemented for CreateUserFcmToken`
            )
        },
        toFirestore: (obj: CreateUserFcmToken): DocumentData => {
            return obj.toJson()
        }
    })

/**
 * Provides a reference to a userFcmToken document for creating.
 * @param userFcmTokenId - The ID of the userFcmToken document to read.
 */
export const createUserFcmTokenDocumentReference = ({
    userFcmTokenId
}: {
    userFcmTokenId: string
}): DocumentReference<CreateUserFcmToken> =>
    createUserFcmTokenCollectionReference.doc(userFcmTokenId)

/**
 * Provides a reference to the userFcmTokens collection for updating.
 */
export const updateUserFcmTokenCollectionReference = db
    .collection(`userFcmTokens`)
    .withConverter<UpdateUserFcmToken>({
        fromFirestore: () => {
            throw new Error(
                `fromFirestore is not implemented for UpdateUserFcmToken`
            )
        },
        toFirestore: (obj: UpdateUserFcmToken): DocumentData => {
            return obj.toJson()
        }
    })

/**
 * Provides a reference to a userFcmToken document for updating.
 * @param userFcmTokenId - The ID of the userFcmToken document to read.
 */
export const updateUserFcmTokenDocumentReference = ({
    userFcmTokenId
}: {
    userFcmTokenId: string
}): DocumentReference<UpdateUserFcmToken> =>
    updateUserFcmTokenCollectionReference.doc(userFcmTokenId)

/**
 * Provides a reference to the userFcmTokens collection for deleting.
 */
export const deleteUserFcmTokenCollectionReference = db
    .collection(`userFcmTokens`)
    .withConverter<DeleteUserFcmToken>({
        fromFirestore: () => {
            throw new Error(
                `fromFirestore is not implemented for DeleteUserFcmToken`
            )
        },
        toFirestore: (): DocumentData => {
            throw new Error(
                `toFirestore is not implemented for DeleteUserFcmToken`
            )
        }
    })

/**
 * Provides a reference to a userFcmToken document for deleting.
 * @param userFcmTokenId - The ID of the userFcmToken document to read.
 */
export const deleteUserFcmTokenDocumentReference = ({
    userFcmTokenId
}: {
    userFcmTokenId: string
}): DocumentReference<DeleteUserFcmToken> =>
    deleteUserFcmTokenCollectionReference.doc(userFcmTokenId)

/**
 * Manages queries against the userFcmTokens collection.
 */
export class UserFcmTokenQuery {
    /**
     * Fetches userFcmToken documents.
     * @param queryBuilder - Function to modify the query.
     * @param compare - Function to sort the results.
     */
    async fetchDocuments({
        queryBuilder,
        compare
    }: {
        queryBuilder?: (
            query: Query<ReadUserFcmToken>
        ) => Query<ReadUserFcmToken>
        compare?: (lhs: ReadUserFcmToken, rhs: ReadUserFcmToken) => number
    }): Promise<ReadUserFcmToken[]> {
        let query: Query<ReadUserFcmToken> = readUserFcmTokenCollectionReference
        if (queryBuilder != undefined) {
            query = queryBuilder(query)
        }
        const qs: QuerySnapshot<ReadUserFcmToken> = await query.get()
        let result = qs.docs.map(
            (qds: QueryDocumentSnapshot<ReadUserFcmToken>) => qds.data()
        )
        if (compare != undefined) {
            result = result.sort(compare)
        }
        return result
    }

    /**
     * Fetches a specific userFcmToken document.
     * @param userFcmTokenId - The ID of the userFcmToken document to fetch.
     */
    async fetchDocument({
        userFcmTokenId
    }: {
        userFcmTokenId: string
    }): Promise<ReadUserFcmToken | undefined> {
        const ds = await readUserFcmTokenDocumentReference({
            userFcmTokenId
        }).get()
        return ds.data()
    }

    /**
     * Adds a userFcmToken document.
     * @param createUserFcmToken - The userFcmToken details to add.
     */
    async add({
        createUserFcmToken
    }: {
        createUserFcmToken: CreateUserFcmToken
    }): Promise<DocumentReference<CreateUserFcmToken>> {
        return createUserFcmTokenCollectionReference.add(createUserFcmToken)
    }

    /**
     * Sets a userFcmToken document.
     * @param userFcmTokenId - The ID of the userFcmToken document to set.
     * @param createUserFcmToken - The userFcmToken details to set.
     * @param options - Options for the set operation.
     */
    async set({
        userFcmTokenId,
        createUserFcmToken,
        options
    }: {
        userFcmTokenId: string
        createUserFcmToken: CreateUserFcmToken
        options?: SetOptions
    }): Promise<WriteResult> {
        if (options == undefined) {
            return createUserFcmTokenDocumentReference({
                userFcmTokenId
            }).set(createUserFcmToken)
        } else {
            return createUserFcmTokenDocumentReference({
                userFcmTokenId
            }).set(createUserFcmToken, options)
        }
    }

    /**
     * Updates a specific userFcmToken document.
     * @param userFcmTokenId - The ID of the userFcmToken document to update.
     * @param updateUserFcmToken - The details for updating the userFcmToken.
     */
    async update({
        userFcmTokenId,
        updateUserFcmToken
    }: {
        userFcmTokenId: string
        updateUserFcmToken: UpdateUserFcmToken
    }): Promise<WriteResult> {
        return updateUserFcmTokenDocumentReference({
            userFcmTokenId
        }).update(updateUserFcmToken.toJson())
    }

    /**
     * Deletes a specific userFcmToken document.
     * @param userFcmTokenId - The ID of the userFcmToken document to delete.
     */
    async delete({
        userFcmTokenId
    }: {
        userFcmTokenId: string
    }): Promise<WriteResult> {
        return deleteUserFcmTokenDocumentReference({
            userFcmTokenId
        }).delete()
    }
}
